#39 eager selects

gbadner opened the pull request:
PR for https://github.com/hibernate/hibernate-reactive/issues/39.

Depends on hibernate-orm PR for [HHH-14075](https://hibernate.atlassian.net/browse/HHH-14075):
https://github.com/hibernate/hibernate-orm/pull/3436

gbadner commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveEntityIdentityInsertAction.java:
```diff
@@ -45,19 +44,20 @@ public void execute() throws HibernateException {
 
 	@Override
 	public CompletionStage<Void> reactiveExecute() throws HibernateException {
```
The change to this method was necessary to ensure that identity ID values were obtained before completing `persist`.

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/ReactiveLoaderBasedLoader.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * An interface intended for {@link ReactiveLoader} implementations that also
+ * extend a {@link org.hibernate.loader.Loader} implementation.
```
@gbadner do we actually have any `ReactiveLoader`s which aren’t `Loader`s?

gbadner replied:
@gavinking , yes, `ReactivePlanEntityLoader` implements `ReactiveLoader`, but does not extend `Loader`.

gavinking replied:
Ah OK of course, I forgot about that one :)

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveLoaderBasedResultSetProcessor.java:
```diff
@@ -0,0 +1,187 @@
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A {@link ReactiveResultSetProcessor} intended to be used by {@link ReactiveLoaderBasedLoader} implementations.
```
Does the logic in this class come from some specific place in Hibernate core?

gbadner replied:
I've added Javadoc indicating the methods upon which they are based.

gavinking replied:
Great, thanks

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveResultSetProcessor.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+public interface ReactiveResultSetProcessor {
```
Is this supposed to be a reactive version of `ResultSetProcessor` in core or does it have a different purpose?

gbadner replied:
hibernate-orm has a `ResultSetProcessor` that is used by `LoadPlan`-based loaders. 

I added `ReactiveResultSetProcessor` to unify how `ReactiveLoader#doReactiveQueryAndInitializeNonLazyCollections` processes a `ResultSet`s  by  both `Loader`-based and `LoadPlan`-based loaders.

gavinking replied:
OK fine, so it’s something new.

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/entity/impl/ReactivePlanEntityLoader.java:
```diff
@@ -236,4 +257,173 @@ public String preprocessSQL(String sql,
 		return sql;
 	}
+
+	private static class ReactiveLoadPlanBasedResultSetProcessor extends ResultSetProcessorImpl
```
Same question here: does the logic herein correspond to something specific in core?

gbadner replied:
I've added Javadoc about this.

gbadner commented:
@gavinking, I've pushed a commit that addresses your questions:
https://github.com/hibernate/hibernate-reactive/pull/259/commits/a4ce598dee31339f1fbe13a716bae78e1933d575

Does that look ok?

gavinking commented:
> @gavinking, I've pushed a commit that addresses your questions:
> 
> https://github.com/hibernate/hibernate-reactive/pull/259/commits/a4ce598dee31339f1fbe13a716bae78e1933d575
> 
> 
> 
> Does that look ok?

Perfect, that’s great, thanks.

gavinking commented:
Alright, so I guess we have to wait for the next release of core to be able to merge this. That’s a bit of a bummer.

gbadner commented:
@gavinking, I just volunteered to release ORM 5.4.18 next week...

Do you have any suggestions about what I should work on next?

gavinking commented:
Well, there’s a couple of things that I was thinking about: 

- @ManyToMany associations would be great to have, but I guess it’s good bit of work.
- Easier, probably, would be batch inserts, which would help with benchmarks.

Thoughts?

gavinking commented:
> @gavinking, I just volunteered to release ORM 5.4.18 next week...

OK, that’s great.

gbadner commented:
@gavinking, I assigned the "batch inserts (and updates)" issue to myself.

gavinking commented:
@gbadner this seems to be working great, but one thing I noticed was if I change it to use:

    @Fetch(FetchMode.SUBSELECT)

at `FetchModeSelectEagerTest:177`, I get the failure below.

I wondered if it might just be something very simple (since IIRC subselect fetching really works much the same as regular eager select fetching). But nothing obvious jumps out at me.

Any ideas?

```
19046 DEBUG org.hibernate.SQL  - select fetchmodes0_.id as id1_1_, fetchmodes0_.parent_id as parent_i4_1_, fetchmodes0_.string as string2_1_, fetchmodes0_.version as version3_1_ from Node fetchmodes0_ order by fetchmodes0_.id
19054 DEBUG org.hibernate.SQL  - select elements0_.node_id as node_id2_0_1_, elements0_.id as id1_0_1_, elements0_.id as id1_0_0_, elements0_.node_id as node_id2_0_0_ from Element elements0_ where elements0_.node_id in (select fetchmodes0_.id from Node fetchmodes0_ )
19068 ERROR org.hibernate.reactive.errors  - failed to execute statement [select fetchmodes0_.id as id1_1_, fetchmodes0_.parent_id as parent_i4_1_, fetchmodes0_.string as string2_1_, fetchmodes0_.version as version3_1_ from Node fetchmodes0_ order by fetchmodes0_.id]
19068 ERROR org.hibernate.reactive.errors  - could not execute query
java.util.concurrent.CompletionException: java.lang.AssertionError: Shouldn't be here
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1005)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fire(ReactiveSessionImpl.java:1038)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveInitializeCollection(ReactiveSessionImpl.java:382)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveFetch(ReactiveSessionImpl.java:171)
	at org.hibernate.reactive.engine.impl.ReactivePersistenceContextAdapter$NonLazyCollectionInitializer.lambda$accept$1(ReactivePersistenceContextAdapter.java:56)
	at org.hibernate.reactive.engine.impl.ReactivePersistenceContextAdapter$NonLazyCollectionInitializer$$Lambda$361/0000000000000000.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at org.hibernate.reactive.pool.impl.Handlers.lambda$toCompletionStage$0(Handlers.java:26)
	at org.hibernate.reactive.pool.impl.Handlers$$Lambda$64/0000000000000000.handle(Unknown Source)
	at io.vertx.sqlclient.impl.SqlResultHandler.complete(SqlResultHandler.java:97)
	at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:86)
	at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:33)
	at io.vertx.sqlclient.impl.SqlConnectionImpl.lambda$schedule$0(SqlConnectionImpl.java:52)
	at io.vertx.sqlclient.impl.SqlConnectionImpl$$Lambda$211/0000000000000000.handle(Unknown Source)
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:188)
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:78)
	at io.vertx.sqlclient.impl.SocketConnectionBase$$Lambda$201/0000000000000000.handle(Unknown Source)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$2(NetSocketImpl.java:101)
	at io.vertx.core.net.impl.NetSocketImpl$$Lambda$192/0000000000000000.handle(Unknown Source)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:237)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:127)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:357)
	at io.vertx.core.net.impl.VertxHandler$$Lambda$193/0000000000000000.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.vertx.pgclient.impl.codec.PgEncoder.lambda$write$0(PgEncoder.java:78)
	at io.vertx.pgclient.impl.codec.PgEncoder$$Lambda$202/0000000000000000.handle(Unknown Source)
	at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:136)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:227)
	at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:86)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:821)
Caused by: java.lang.AssertionError: Shouldn't be here
	at org.hibernate.reactive.event.impl.DefaultReactiveInitializeCollectionEventListener.onReactiveInitializeCollection(DefaultReactiveInitializeCollectionEventListener.java:84)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$$Lambda$363/0000000000000000.apply(Unknown Source)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.lambda$fire$39(ReactiveSessionImpl.java:1038)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$$Lambda$227/0000000000000000.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
	... 57 more

java.lang.AssertionError: Shouldn't be here
java.util.concurrent.CompletionException: java.lang.AssertionError: Shouldn't be here
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1005)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fire(ReactiveSessionImpl.java:1038)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveInitializeCollection(ReactiveSessionImpl.java:382)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveFetch(ReactiveSessionImpl.java:171)
	at org.hibernate.reactive.engine.impl.ReactivePersistenceContextAdapter$NonLazyCollectionInitializer.lambda$accept$1(ReactivePersistenceContextAdapter.java:56)
	at org.hibernate.reactive.engine.impl.ReactivePersistenceContextAdapter$NonLazyCollectionInitializer$$Lambda$361/0000000000000000.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at org.hibernate.reactive.pool.impl.Handlers.lambda$toCompletionStage$0(Handlers.java:26)
	at org.hibernate.reactive.pool.impl.Handlers$$Lambda$64/0000000000000000.handle(Unknown Source)
	at io.vertx.sqlclient.impl.SqlResultHandler.complete(SqlResultHandler.java:97)
	at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:86)
	at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:33)
	at io.vertx.sqlclient.impl.SqlConnectionImpl.lambda$schedule$0(SqlConnectionImpl.java:52)
	at io.vertx.sqlclient.impl.SqlConnectionImpl$$Lambda$211/0000000000000000.handle(Unknown Source)
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:188)
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:78)
	at io.vertx.sqlclient.impl.SocketConnectionBase$$Lambda$201/0000000000000000.handle(Unknown Source)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$2(NetSocketImpl.java:101)
	at io.vertx.core.net.impl.NetSocketImpl$$Lambda$192/0000000000000000.handle(Unknown Source)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:237)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:127)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:357)
	at io.vertx.core.net.impl.VertxHandler$$Lambda$193/0000000000000000.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.vertx.pgclient.impl.codec.PgEncoder.lambda$write$0(PgEncoder.java:78)
	at io.vertx.pgclient.impl.codec.PgEncoder$$Lambda$202/0000000000000000.handle(Unknown Source)
	at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:136)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:227)
	at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:86)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:821)
Caused by: java.lang.AssertionError: Shouldn't be here
	at org.hibernate.reactive.event.impl.DefaultReactiveInitializeCollectionEventListener.onReactiveInitializeCollection(DefaultReactiveInitializeCollectionEventListener.java:84)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$$Lambda$363/0000000000000000.apply(Unknown Source)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.lambda$fire$39(ReactiveSessionImpl.java:1038)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$$Lambda$227/0000000000000000.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
	... 57 more
```

Sanne commented:
@gavinking we're preparing an ORM release to happen this afternoon, to include the prerequisite patch @gbadner created for this improvement.   If you know of other PRs you need, please let me know?

gavinking commented:
I can’t think of anything else.

gavinking commented on hibernate-reactive-core/src/test/java/org/hibernate/reactive/FetchModeSelectEagerTest.java:
```diff
@@ -0,0 +1,250 @@
+import java.util.List;
+import java.util.Objects;
+
+public class FetchModeSelectEagerTest extends BaseReactiveTest {
```
One more minor thing, @gbadner, I always actually intended for `EagerTest` to be changed to test select fetching for the parent as well. So I don't think we need `FetchModeSelectEagerTest`, we just need to change the fetch mode of `Element.node`, uncommenting the commented line.

gbadner replied:
@gavinking , OK, I'll change `EagerTest` and remove `FetchModeSelectEagerTest`.

gbadner replied:
Actually, I'll change `FetchModeSelectEagerTest` to `FetchModeSubselectEagerTest` and look into why this is failing for `FetchMode.SUBSELECT`.

gavinking replied:
Awesome, thanks.

Sanne commented:
Depends on #262

gbadner commented:
@gavinking, I merged Sanne's [PR](https://github.com/hibernate/hibernate-reactive/pull/262) that upgrades ORM to 5.4.18.Final, and rebased this PR, and added a [commit](https://github.com/hibernate/hibernate-reactive/pull/259/commits/d5e0b5ceee4ba21cc72d68ea2086ee276c647028) with the change to EagerTest and a couple of miscellaneous changes to prepare this to be pushed upstream.

IIUC, this should be good to go.

I will continue working on FetchMode.SUBSELECT using https://github.com/hibernate/hibernate-reactive/issues/261

gavinking commented:
Perfect, thanks so much @gbadner it's really great to have removed this limitation.

Sanne commented:
agreed! thanks.

@gavinking : I already upgraded Quarkus to latest ORM. We could do a Reactive release.. want me to trigger one now, or is there something else you'd want to see included in the next?

we're 1 day away from Quarkus code freeze for **1.6 CR1** . I expect we can include more fixes before .Final but I'd prefer to see as much as possible included for CR1.

gavinking commented:
@Sanne yeah, I think it definitely makes sense to do a release
